junit-plugin/src/main/scala/org/scalajs/junit/plugin/ScalaJSJUnitPlugin.scala

```diff
@@ -0,0 +1,447 @@
+package org.scalajs.junit.plugin
+
+import scala.language.reflectiveCalls
+
+import org.scalajs.junit._
+
+import scala.reflect.internal.Flags
+import scala.tools.nsc._
+import scala.tools.nsc.plugins.{
+Plugin => NscPlugin, PluginComponent => NscPluginComponent
+}
+
+class ScalaJSJUnitPlugin(val global: Global) extends NscPlugin {
+  selfPlugin =>
+
+  val name: String = "Scala.js JUnit plugin"
+
+  val components: List[NscPluginComponent] =
+    List(ScalaJSJUnitPluginComponent)
+
+  val description: String = "Adapts JUnit test classes to Scala.js"
+
+  object ScalaJSJUnitPluginComponent extends plugins.PluginComponent
+      with transform.Transform {
+
+    protected def newTransformer(
+        unit: global.CompilationUnit): global.Transformer = {
+      new ScalaJSJUnitPluginTransformer
+    }
+
+    val global: Global = selfPlugin.global
+    val phaseName: String = "junit-inject"
+    val runsAfter: List[String] = List("mixin")
+    override val runsBefore: List[String] = List("jscode")
+
+    class ScalaJSJUnitPluginTransformer extends global.Transformer {
+
+      import global._
+
+      type ScalaJSPlugin = NscPlugin {
+        def PrepInteropComponent: AnyRef {
+          def genExportMember(sym: Symbol): List[Tree]
+          def registerModuleExports(sym: Symbol): Unit
+          def genNamedExport(defSym: Symbol, jsName: String, pos: Position): Unit
+        }
+      }
+
+      lazy val scalaJSPlugin = {
+        global.plugins.collectFirst {
+          case pl if pl.getClass.getName == "org.scalajs.core.compiler.ScalaJSPlugin" => pl
+        }.get.asInstanceOf[ScalaJSPlugin]
+      }
+
+      override def transform(tree: Tree): Tree = tree match {
+        case tree: PackageDef =>
+          def classHasJUnitAnnotations(clazz: ClassDef): Boolean = {
+            clazz.impl.body.exists {
+              case ddef: DefDef => hasAnnotation[org.junit.Test](ddef)
+              case _ => false
+            }
+          }
+          val newStats = tree.stats.groupBy {
+            case clDef: ClassDef => clDef.name
+            case _ => "not a class def"
```

Comment:
In theory, there could exist a class whose name is literally `not a class def`.
Maybe `Some(clDef.name)` versus `None`?